### PR TITLE
`readonly`-ify fields based on whole program view

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
Marking fields `readonly` has a couple benefits:

* Codegen can assume the value doesn't change and can inline the values into code.
* In the static constructor interpreter, we allow reading readonly fields of other classes. So `class A { public static int X = 123 } class B { static int Y = A.X }` - `A` can be preinitialized at compile time, but `B` cannot. However: `class A { public static readonly int X = 123 } class B { static int Y = A.X }` both `A` and `B` can preinitialize at compile time.

In this PR, I'm adding infrastructure to collect information about field writes during whole program analysis. This collects information about fields that are effectively read-only. Then during compilation, we use this data instead of the `IsInitOnly` metadata. The big advantage of this approach (as opposed to e.g. a Roslyn analyzer) is that we look at the app after trimming - some fields might be writable, but if we discarded the code that writes them, they're read only.

This is also a correctness fix - the compiler would previously assume preinitializing instances of classes with all-readonly instance fields is safe, but one can set them with reflection and this would lead to a crash in the write barrier. Not that anyone should be reflection-setting readonly fields outside of deserialization scenarios.

Cc @dotnet/ilc-contrib

### CODE REVIEW

File: src/coreclr/tools/aot/ILCompiler.Compiler/IL/ILImporter.Scanner.cs [hunk 1]
```diff
@@ -974,13 +974,36 @@ private void ImportReadOnlyPrefix()
             _isReadOnly = true;
         }
 
-        private void ImportFieldAccess(int token, bool isStatic, string reason)
+        private void ImportFieldAccess(int token, bool isStatic, bool? write, string reason)
```
Comment (SingleAccretion) [inline src/coreclr/tools/aot/ILCompiler.Compiler/IL/ILImporter.Scanner.cs:977]: Does this analysis interact correctly with unsafe code?

E. g. `&s + <offset> = <value>;`

Comment (MichalStrehovsky) [reply src/coreclr/tools/aot/ILCompiler.Compiler/IL/ILImporter.Scanner.cs:977]: It doesn't, and neither with interop.

Comment (jkotas) [reply src/coreclr/tools/aot/ILCompiler.Compiler/IL/ILImporter.Scanner.cs:977]: This should not be a problem in practice, given what we use the result of the analysis for today. Correct?

Comment (MichalStrehovsky) [reply src/coreclr/tools/aot/ILCompiler.Compiler/IL/ILImporter.Scanner.cs:977]: After 865875849d186d08a071f62aaa84c51cf30f966a I can't think of a problematic case - this should not be making any new assumptions.

File: src/tests/nativeaot/SmokeTests/TrimmingBehaviors/DeadCodeElimination.cs [hunk 2]
```diff
@@ -378,6 +380,100 @@ public static void Run()
         }
     }
```
[the remark this answers is not in the document]
Comment (MichalStrehovsky) [reply src/tests/nativeaot/SmokeTests/TrimmingBehaviors/DeadCodeElimination.cs:381]: Test added.

File: src/coreclr/tools/aot/ILCompiler.Compiler/IL/ILImporter.Scanner.cs (hunk 1 above)
Comment (SingleAccretion) [reply src/coreclr/tools/aot/ILCompiler.Compiler/IL/ILImporter.Scanner.cs:977]: Are the consumers restricted enough that even if the field is only written to via pointer arithmetic, things still work?

```cs
[StructLayout(LayoutKind.Explicit)]
class Xyz
{
    [FieldOffset(0)]
    public int FieldOne;
    [FieldOffset(4)]
    public int FieldTwo;
}

Xyz xyz = { ... };
Unsafe.AddByteOffset(ref xyz.FieldOne, 4) = { ... };
```

Comment (MichalStrehovsky) [reply src/coreclr/tools/aot/ILCompiler.Compiler/IL/ILImporter.Scanner.cs:977]: We go over all fields on the class and check for read only on each. The check is in the preexisting code that we're switching from believing isliteral to using this analysis (that's the correctness fix part from the pr description).

**Review by jkotas — APPROVED** [review]
Nice!

**Pull request merged**